### PR TITLE
Add a LoRa optimization branch

### PR DESCRIPTION
**Pull request by Todobe**
## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - `/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`
4. After green CI and required approvals, ask Merge Oncalls to merge.

### CODE REVIEW

File: python/sglang/srt/lora/torch_ops/lora_ops.py [hunk 1]
```diff
@@ -17,6 +17,17 @@ def sgemm_lora_a_fwd(
         return torch.zeros(total_seq_len, 0, dtype=inputs.dtype, device=inputs.device)
 
     num_loras, weight_out_dim, _ = weights.shape
+    
+    if torch.all(weight_indices == weight_indices[0]):
```
Comment (liupeng374) [inline python/sglang/srt/lora/torch_ops/lora_ops.py]: This sync might interrupt the overlap-scheduler.

**Comment by vlserov** [comment]
Fuse same LoRA requests in ops is not needed, since it has been made for batch by lines below:
https://github.com/sgl-project/sglang/blob/6586f44ad4f6169162384d9342f80c48a0c9c593/python/sglang/srt/lora/backend/torch_backend.py#L192-L196

File: python/sglang/srt/lora/torch_ops/lora_ops.py [hunk 2]
```diff
@@ -17,6 +18,17 @@ def sgemm_lora_a_fwd(
         return torch.zeros(total_seq_len, 0, dtype=inputs.dtype, device=inputs.device)
 
     num_loras, weight_out_dim, _ = weights.shape
+    
+    if get_global_server_args().disable_overlap_schedule and torch.all(weight_indices == weight_indices[0]):
+        idx = weight_indices[0]
+        rank = lora_ranks[idx]
+        scaling = scaling_tensor[idx]
+        if rank > 0:
+            output = torch.einsum("si, oi -> so", inputs, weights[idx, :, :]) * scaling
```
Comment (vlserov) [inline python/sglang/srt/lora/torch_ops/lora_ops.py:27]: In case pinned adapters scaling will be non-one-element tensor, which cause exception

**Review by vlserov — CHANGES_REQUESTED** [review]
Changes is not needed since equals LoRA request already have been merged.
Fuse same LoRA requests in ops is not needed, since it has been made for batch by lines below:

https://github.com/sgl-project/sglang/blob/6586f44ad4f6169162384d9342f80c48a0c9c593/python/sglang/srt/lora/backend/torch_backend.py#L192-L196

File: python/sglang/srt/lora/torch_ops/lora_ops.py [hunk 3]
```diff
@@ -2,6 +2,7 @@
 
 import torch
 
+from sglang.srt.server_args import get_global_server_args
```
Comment (vlserov) [inline python/sglang/srt/lora/torch_ops/lora_ops.py:5]: No need

File: python/sglang/srt/lora/torch_ops/lora_ops.py [hunk 4]
```diff
@@ -17,6 +18,17 @@ def sgemm_lora_a_fwd(
         return torch.zeros(total_seq_len, 0, dtype=inputs.dtype, device=inputs.device)
 
     num_loras, weight_out_dim, _ = weights.shape
+    
+    if len(weight_indices) == 1:
+        idx = weight_indices[0]
+        rank = lora_ranks[idx]
```
Comment (vlserov) [inline python/sglang/srt/lora/torch_ops/lora_ops.py:24]: Cause accuracy issue in case adapter rank is not equal max_lora_rank
